libc/intrin/g_fds.c

```diff
@@ -43,29 +43,37 @@ textstartup void InitializeFileDescriptors(void) {
   struct Fds *fds;
   __fds_lock_obj._type = PTHREAD_MUTEX_RECURSIVE;
   fds = VEIL("r", &g_fds);
-  fds->p = fds->e = (void *)0x6fe000040000;
   fds->n = 4;
   fds->f = 3;
-  fds->e = _extend(fds->p, fds->n * sizeof(*fds->p), fds->e, 0x6ff000000000);
-  if (IsMetal()) {
-    extern const char vga_console[];
-    pushmov(&fds->f, 3ull);
-    if (weaken(vga_console)) {
-      fds->p[0].kind = pushpop(kFdConsole);
-      fds->p[1].kind = pushpop(kFdConsole);
-      fds->p[2].kind = pushpop(kFdConsole);
-    } else {
-      fds->p[0].kind = pushpop(kFdSerial);
-      fds->p[1].kind = pushpop(kFdSerial);
-      fds->p[2].kind = pushpop(kFdSerial);
-    }
-    fds->p[0].handle = VEIL("r", 0x3F8ull);
-    fds->p[1].handle = VEIL("r", 0x3F8ull);
-    fds->p[2].handle = VEIL("r", 0x3F8ull);
-  } else if (IsWindows()) {
+
+  if (IsWindows()) {
+    // Use old fixed size descriptor array
+    // TODO find way to use dynamic array to prevent potental race condition like on other OSes
+    fds->p = fds->__win_init_p;
+
     SetupWinStd(fds, 0, kNtStdInputHandle);
     SetupWinStd(fds, 1, kNtStdOutputHandle);
     SetupWinStd(fds, 2, kNtStdErrorHandle);
+  } else {
+    fds->p = fds->e = (void *)0x6fe000040000;
+    fds->e = _extend(fds->p, fds->n * sizeof(*fds->p), fds->e, 0x6ff000000000);
```

Comment:
I'm surprised _extend ended up in the vista branch because I didn't intend to have it in there. We might want to recut the branch or simply revert the change that added `_extend`

Reply:
I will update this pull when I have a chance as I have a working version with _extend at [vistafix2](https://github.com/JoshuaWierenga/cosmopolitan/tree/vistafix2). The only issue is it uses memory assigned to asan on windows 8 and older, at least according to [memtrack64](https://github.com/jart/cosmopolitan/blob/master/libc/runtime/memtrack64.txt). Issues to consider with that fix include whether or not the address is actually stored somewhere to avoid dynamically allocating something else to it and if the platform check substantially impacts the performance and binary size.

It would also be useful to have some testing on vista/7/8, even if it's just something I set up locally, how is cross-platform testing handled typically? That would help ensure my fix works and there are no issues in newer commits since I have just been working on top of the _extend commit directly.

Reply:
If the patch doesn't totally work and isn't reasonably fixable, or the performance/size is adversely impacted then I can go back to the previous commit. However, I see no reason to drop functionality that works fine on other platforms if not absolutely necessary.

Reply:
If you want to keep _extend() then all you need to do is just pick a 32-bit address to use when it's Windows 7 or lower. There's a couple text files in libc/runtime/ where you can document your decisions. You should see in memtrack.internal.h the ternary statements used in the `#define` that special case the hard-coded addresses. We just need that to happen for the new code that's using the _extend() API.

Let me know as soon as you think it's ready for merge, and I'll merge it. I can also do testing legwork for you too, where I run your change on my test fleet to make sure there aren't any regressions with other operating systems.

Thanks so much!